site/app/Livewire/Finder.php

```diff
@@ -0,0 +1,530 @@
+    ): void {
+        $message = collect($lines)->values()->flatten()->join('<br />');
+        $this->flashMessage($message, $bsClass);
+    }
+
+    public function openFolder(Folder $folder): void
+    {
+        $this->redirect(route('folders.show', $folder->id));
+    }
+
+    /**
+     * Add or remove a filter (search box, like name, box 2, etc.) on the card.
+     */
+    public function toggleFilterSearchBox(string $filter, bool $checked): void
+    {
+        $success = $this->validateAndFlash(
+            [
+                'type' => $filter,
+            ],
+            [
+                'type' => 'required|string|in:name,'.CardBox::Box2.','.CardBox::Box3.','.CardBox::Box4,
+            ],
+        );
+        if (! $success) {
+            return;
+        }
+
+        $this->filterSearchBoxes[$filter] = $checked;
+    }
+
+    /**
+     * Reinitialise all filters and sort to their default values.
+     */
+    public function clearFiltersAndSort(): void
+    {
+        $this->initFilters();
+
+        $this->sortColumn = static::DEFAULT_SORT_COLUMN;
+        $this->sortDirection = static::DEFAULT_SORT_DIRECTION;
+    }
+
+    /**
+     * Change the sorting column and direction.
+     */
+    public function sort(string $column, string $direction): void
+    {
+        $success = $this->validateAndFlash(
+            [
+                'column' => $column,
+                'direction' => $direction,
+            ],
+            [
+                'column' => 'required|string|in:title,state_name,created_at,editors_list,tags_list,'.static::DEFAULT_SORT_COLUMN,
+                'direction' => 'required|string|in:asc,desc',
+            ],
+        );
+
+        if (! $success) {
+            return;
+        }
+
+        $this->sortColumn = $column;
+        $this->sortDirection = $direction;
+    }
+
+    /**
+     * Clone a card.
+     */
+    public function cloneCard(Card $card): void
+    {
+        $this->authorize('manage', $card);
+
+        (new CloneCardService($card))->clone();
+    }
+
+    /**
+     * Clone a folder. If $displayFlash is true, a flash message will be
+     * displayed when the folder is cloned.
+     */
+    public function cloneFolder(
+        Folder $folder,
+        bool $displayFlash = false,
+    ): void {
+        // We do not check folder's children since it's the same as the parent.
+        $this->authorize('manage', $folder);
+
+        (new CloneFolderService($folder))->clone();
+
+        if ($displayFlash) {
+            $this->flashMessage(trans('courses.finder.menu.copy.success'));
+        }
+    }
+
+    /**
+     * Clone all items in the $keys array ($keys is 'card-1', 'folder-2', etc.).
+     */
+    public function cloneMultiple(array $keys): void
+    {
+        $this->keysToEntities($keys)
+            ->each(fn ($item) => $this->authorize('manage', $item))
+            ->each(fn ($entity) => MassCloneService::getCloneService($entity)->clone());
```

Comment:
We could execute a job when cloning inside another course. But if we clone in the same course, we will not be able to update the view with the new items since we will not have any callback (except if we do polling).

We could do async job when cloning in another course, and execute these job synced (dispatchSync) when cloning in the same course.

However, we will not have any callback that the cloning in another course has succeeded (currently a toast is displayed saying if it was successful or not).

I think it's better for the user to wait for the cloning to be completed and don't think there will be cloning taking longer than a couple of seconds in worst cases. Perhaps showing the user a loader will be a good compromise that its action is still pending and he can't do anything in between (he can but its further actions will be put in wait until the cloning is complete).

Reply:
_Perhaps showing the user a loader will be a good compromise that its action is still pending and he can't do anything in between (he can but its further actions will be put in wait until the cloning is complete)._

👍

Reply:
An indicator will be shown if the request takes longer than 500ms.